pkg/controller/cloudlet_api.go

```diff
@@ -2625,3 +2625,114 @@ func (s *CloudletApi) updateDefaultMultiTenantClusterWorker(ctx context.Context,
 		s.all.clusterInstApi.deleteDefaultMultiTenantCluster(ctx, key)
 	}
 }
+
+func (s *CloudletApi) ChangeCloudletDNS(key *edgeproto.CloudletKey, inCb edgeproto.CloudletApi_ChangeCloudletDNSServer) (reterr error) {
+	ctx := inCb.Context()
+	cctx := DefCallContext()
+
+	cloudlet := edgeproto.Cloudlet{}
+	if !s.store.Get(ctx, key, &cloudlet) {
+		return key.NotFoundError()
+	}
+	if !cloudlet.CrmOnEdge {
+		return fmt.Errorf("unsupported - only cloudlets with crm on edge are currently supported")
+	}
+
+	if cloudlet.MaintenanceState != dme.MaintenanceState_UNDER_MAINTENANCE {
+		return fmt.Errorf("maintenance mode is required to update DNS")
+	}
+
+	streamCb, cb := s.all.streamObjApi.newStream(ctx, cctx, key.StreamKey(), inCb)
+
+	// Step 1 - update rootLb fqdn
+	// Does our current DNS app root name match what the clouldet alredy have?
+	// TODO - why does the streaming not work for cloudlet crm update
+	if strings.HasSuffix(cloudlet.RootLbFqdn, *appDNSRoot) {
+		cb.Send(&edgeproto.Result{Message: "Cloudlet rootLB is already set correctly"})
+		log.SpanLog(ctx, log.DebugLevelApi, "Current cloudlet fqdn already contains appDNSRoot suffix - nothing to do")
+	} else {
+		log.SpanLog(ctx, log.DebugLevelApi, "Update rootLB fqdn", "old", cloudlet.RootLbFqdn, "new", getCloudletRootLBFQDN(&cloudlet))
+		cb.Send(&edgeproto.Result{Message: "Updating Cloudlet RootLb FQDN"})
```

Comment:
I thought about it, and I'd rather have the cloudlet in maintenance for the duration of the api - this way it eliminates possible race-conditions with appInsts, clusterInsts being modified.

Reply:
Sounds good. The scenario I was thinking about is that since we are first collecting clusterInsts and only after that updating them one-by-one, there is a moment between these two actions that a new cluster could be created.  But in any case - I'll keep maintenance requirement for the whole api work.